docs/blueprints/slow-down-and-retry.md

```diff
@@ -0,0 +1,87 @@
+# Slow Down and Retry
+
+## Abstract
+
+Our goals are to:
+
+* Improve the user experience when the request limit is exceeded by allowing
+  clients to gracefully recover
+* Avoid duplication of the implemention work for slow down and retry logic
+
+## Background
+
+The satellite has a rate limit on API requests per second per project. When a
+client exceeds the rate limit it is sent a `Too Many Requests` error. This kind
+of error is a retriable, but requires the client to also implement some form of
+backoff (e.g. exponential).
+
+Without a retry this error becomes fatal and can cause operations to fail that
+would otherwise succeed if performed at a slower pace. For example, long
+listing operations that proceed too quickly can fail and have to be restarted
+(e.g. observed often in the S3 gateway).
+
+In particular large uploads are badly impacted because a single `Too Many
+Requests` error would result in the entire upload having to be redone. Uploads
+are also a compelling reason for putting the logic in libuplink itself because
+only libuplink has the details necessary to retry at the appropriate level
+(e.g. segment or node request).
+
+## Design
+
+Libuplink internally should be able to handle retry for:
+* every request to Metainfo API when error `Too Many Requests` is returned
+* segment upload request (several requests combined)
+* segment download request (several requests combined)
+
+For most of Metainfo requests and `Too Many Requests` error we need simply apply retry and slow down logic without any additional constraints.
+
+Special case to handle is Metainfo API `Batch` request which is currently used only during upload. This request takes a list of requests to execute on satellite side and returns responses from each of it. If one of requests will fail then on retry we should continue from first not executed request to finish whole operation. It's necessary because `Batch` request doesn't provide any kind of rollback in case of error while executing list of requests and we can be in a state where retrying from scratch won't be possible.
+
```

Comment:
This is missing implicit argument forwarding.

Reply:
I'm not following.

Reply:
Currently we support as a single batch:
```
create object
make inline segment
commit object
```

The "stream id" is carried forward implicitly from create object into the batch items following it. So, for example, when "make inline segment" fails due to too many requests then the client must manually read the "stream id" from "create object" result and then add them to the requests that are retried.